### PR TITLE
[WIP] Separate kernel caching from the interface

### PR DESCRIPTION
**Pull request by theartful**
This is done by creating another class `CachedKernel` that is composed of the original kernel, in addition to the caching data/methods.

### CODE REVIEW

File: src/shogun/kernel/CachedKernel.h [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+#include <shogun/io/SGIO.h>
+#include <shogun/kernel/Kernel.h>
+#include <shogun/lib/common.h>
+#include <shogun/lib/config.h>
+
+namespace shogun
+{
+
+#ifdef USE_SVMLIGHT
```
Comment (karlnapf) [inline src/shogun/kernel/CachedKernel.h:11]: does it use SVMLight?

File: src/shogun/kernel/CachedKernel.h [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+			/** start */
+			int32_t start;
+			/** end */
+			int32_t end;
+			/** of vectors */
+			int32_t num_vectors;
+		};
+#endif // DOXYGEN_SHOULD_SKIP_THIS
+
+		static void* cache_multiple_kernel_row_helper(void* p);
```
Comment (karlnapf) [inline src/shogun/kernel/CachedKernel.h:228]: I assume all this is basically cut/pasted from somewhere else?

Comment (ghost) [reply src/shogun/kernel/CachedKernel.h:228]: yes, from Kernel.h

File: src/shogun/kernel/CachedKernel.h (hunk 1 above)
Comment (ghost) [reply src/shogun/kernel/CachedKernel.h:11]: It's only used by SVMLight

**Comment by iglesias** [comment]
What's the status here, @theartful?

**Comment by ghost** [comment]
Almost all SVMLight related examples are ported, and I added a unit-test that checks that cache-size doesn't affect results.

File: tests/unit/classifier/svm/SVMLight_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+#include <shogun/kernel/GaussianKernel.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/config.h>
+
+using namespace shogun;
+
+#ifdef USE_SVMLIGHT
+TEST(SVMLight, CacheSizeInvariance)
+{
+	const int32_t num_samples = 50;
```
Comment (iglesias) [inline tests/unit/classifier/svm/SVMLight_unittest.cc:15]: Minor on style, uniform initialization is preferred. No need to update it here; I find it a convenient rule.

**Comment by karlnapf** [comment]
I think this can be merged soon?

**Comment by ghost** [comment]
While we're at it, one last thing. I don't like this line here:
```
gaussian_kernel->set_cache_size(cache_size);
```
There are multiple algorithms that use this cache size property, but shouldn't it be a parameter of these algorithms?

**Comment by iglesias** [comment]
> While we're at it, one last thing. I don't like this line here:
> 
> ```
> gaussian_kernel->set_cache_size(cache_size);
> ```
> 
> There are multiple algorithms that use this cache size property, but shouldn't it be a parameter of these algorithms?

It could make better sense, indeed. Why do you think it'd better as parameter?

**Comment by ghost** [comment]
@iglesias It's currently a parameter of the Kernel interface, but if we're going stateless, it shouldn't. Also I think this doesn't concern/affect the behaviour of the kernel as it does the algorithms.

**Comment by gf712** [comment]
@theartful I'll keep this alive with a comment and then can see

**Comment by ghost** [comment]
@gf712 Thanks! I will tie up some of these PRs this weekend.

**Comment by karlnapf** [comment]
@theartful would be great to get this in! :)

**Comment by theartful** [comment]
September first I will be free :") Will look into it then along side my other PRs.

**Comment by karlnapf** [comment]
That'd be ace!

**Pull request closed without merging**